src/nomad_measurements/utils.py

```diff
@@ -153,3 +160,304 @@ def get_bounding_range_2d(ax1, ax2):
+        against the `valid_dataset_paths` if provided before adding the data.
+
+        Args:
+            path (str): The dataset path to be used in the HDF5 file.
+            data (Any): The data to be stored in the HDF5 file.
+            archive_path (str): The path of the quantity in the archive.
+            internal_reference (bool): If True, an internal reference is set to an
+                existing HDF5 dataset.
+            validate_path (bool): If True, the path is validated against the
+                `valid_dataset_paths`.
+            lazy (bool): If True, the file is not written immediately.
+        """
+        if not path:
+            self.logger.warning('HDF5 `path` must be provided.')
+            return
+
+        if validate_path and self.valid_dataset_paths:
+            if path not in self.valid_dataset_paths:
+                self.logger.warning(f'Invalid dataset path "{path}".')
+                return
+
+        dataset = dict(
+            data=data,
+            attrs={},
+            hdf5_path=(
+                f'/uploads/{self.archive.m_context.upload_id}/raw'
+                f'/{self.data_file}#{path}'
+            ),
+            archive_path=archive_path,
+            internal_reference=internal_reference,
+        )
+        # handle the pint.Quantity and add data
+        if isinstance(data, pint.Quantity):
+            dataset['data'] = data.magnitude
+            dataset['attrs'].update({'units': str(data.units)})
+
+        self.hdf5_datasets[path] = dataset
+
+        if not lazy:
+            self.write_file()
+
+    def add_attribute(
+        self,
+        path: str,
+        attrs: dict,
+        lazy: bool = True,
+    ):
+        """
+        Add an attribute to the dataset or group at the given path. The attribute is
+        written lazily (default) when either `read_dataset` or `write_file` method is
+        called.
+
+        Args:
+            path (str): The dataset or group path in the HDF5 file.
+            attrs (dict): The attributes to be added.
+            lazy (bool): If True, the file is not written immediately.
+        """
+        self.hdf5_attributes[path] = attrs
+
+        if not lazy:
+            self.write_file()
+
+    def read_dataset(self, path: str):
+        """
+        Returns the dataset at the given path. If the quantity has `units` as an
+        attribute, tries to returns a `pint.Quantity`. Before returning the dataset, the
+        method also writes the file with any pending datasets.
+
+        Args:
+            path (str): The dataset path in the HDF5 file.
+        """
+        if self.hdf5_datasets or self.hdf5_attributes:
+            self.write_file()
+        if path is None:
+            return
+        file_path, dataset_path = path.split('#')
+        file_name = file_path.rsplit('/raw/', 1)[1]
+        with self.archive.m_context.raw_file(file_name, 'r') as h5file:
```

Comment:
i suspect this would break in case the uplod gets published, can you try it?

Reply:
i will take a look

Reply:
It works fine. I tested by publishing the upload and navigating into the quantities, the references and plotting worked fine. 
Also, once published, these methods will not be able to run anyway because re-processing is not possible.